dart/biomechanics/IKInitializer.cpp

```diff
@@ -0,0 +1,3080 @@
+  if (solver.info() != Eigen::Success)
+    return false;
+
+  // Check if all eigenvalues are positive
+  if ((solver.eigenvalues().array() > 0).all())
+    return true;
+  else
+  {
+    std::cout << "Got non-positive eigenvalues: " << solver.eigenvalues()
+              << std::endl;
+    return false;
+  }
+}
+
+//==============================================================================
+Eigen::Vector3s ensureOnSameSideOfPlane(
+    const std::vector<Eigen::Vector3s>& neutralPoints,
+    Eigen::Vector3s neutralGoal,
+    std::vector<Eigen::Vector3s> actualPoints,
+    Eigen::Vector3s ambiguousReconstructionFromActualPoints)
+{
+  if (neutralPoints.size() < 3 || actualPoints.size() < 3)
+  {
+    return ambiguousReconstructionFromActualPoints;
+  }
+
+  // Choose the first three points and form two vectors
+  const Eigen::Vector3s neutralNormal
+      = ((neutralPoints[1] - neutralPoints[0])
+             .cross(neutralPoints[2] - neutralPoints[0]))
+            .normalized();
+  s_t neutralGoalDistanceFromNormal
+      = (neutralGoal - neutralPoints[0]).dot(neutralNormal);
+
+  const Eigen::Vector3s actualNormal
+      = ((actualPoints[1] - actualPoints[0])
+             .cross(actualPoints[2] - actualPoints[0]))
+            .normalized();
+  s_t ambiguousReconstructionDistanceFromNormal
+      = (ambiguousReconstructionFromActualPoints - actualPoints[0])
+            .dot(actualNormal);
+
+  if (neutralGoalDistanceFromNormal * ambiguousReconstructionDistanceFromNormal
+      < 0)
+  {
+    // If they're on opposite sides of the plane, flip the `actualGoal` to be on
+    // the same side of the plane as the `neutralGoal`.
+    return ambiguousReconstructionFromActualPoints
+           - 2 * ambiguousReconstructionDistanceFromNormal * actualNormal;
+  }
+  else
+  {
+    // If they're already on the same side of the plane, don't worry
+    return ambiguousReconstructionFromActualPoints;
+  }
+}
+
+//==============================================================================
+IKInitializer::IKInitializer(
+    std::shared_ptr<dynamics::Skeleton> skel,
+    std::map<std::string, std::pair<dynamics::BodyNode*, Eigen::Vector3s>>
+        markers,
+    std::vector<std::map<std::string, Eigen::Vector3s>> markerObservations,
+    s_t modelHeightM)
+  : mSkel(skel),
+    mMarkerObservations(markerObservations),
+    mModelHeightM(modelHeightM)
+{
+  // 1. Convert the marker map to an ordered list
+  for (auto& pair : markers)
+  {
+    mMarkerNames.push_back(pair.first);
+    mMarkers.push_back(pair.second);
+  }
+
+  Eigen::VectorXs oldPositions = mSkel->getPositions();
+  mSkel->setPositions(Eigen::VectorXs::Zero(mSkel->getNumDofs()));
+  Eigen::VectorXs oldScales = mSkel->getBodyScales();
+  if (modelHeightM > 0)
+  {
+    mSkel->setBodyScales(Eigen::VectorXs::Ones(mSkel->getNumBodyNodes() * 3));
+    s_t defaultHeight
+        = mSkel->getHeight(Eigen::VectorXs::Zero(mSkel->getNumDofs()));
+    if (defaultHeight > 0)
```

Comment:
The height never goes negative, but it can be 0 if there are no meshes loaded along with the skeleton. This is just trying to avoid crashing in that case.